File: leetcode/two_sum.php
<?php
/**
 * @desc: 两数之和
 * @date: 2019-03-20 23:52
 * @author: junhui.zhang
 * 给定一个整数数组 nums 和一个目标值 target，请你在该数组中找出和为目标值的那 两个 整数，并返回他们的数组下标。
 * 你可以假设每种输入只会对应一个答案。但是，你不能重复利用这个数组中同样的元素。
 * 示例:
 * 给定 nums = [2, 7, 11, 15], target = 9
 * 因为 nums[0] + nums[1] = 2 + 7 = 9
 * 所以返回 [0, 1]
 */

class Solution {

    /**
     * @param Integer[] $nums
     * @param Integer $target
     * @return Integer[]
     */
    function twoSum($nums, $target) {
        $count = count($nums);
        $result = array();
        for($i = 0; $i < $count; $i++){
            for($j = $i + 1; $j < $count; $j++){
                if($nums[$i] + $nums[$j] == $target){
                    $result = array($i, $j);
		    return $result;
                }
            }

        }
        
        return ;
    }

    function twoSum2($nums, $target){
        $count = count($nums);
	for($i = 0; $i < $count; $i++){
	    $needed = $target - $nums[$i];
	    if(in_array($needed, $nums)){
		$index = array_search($needed, $nums);
		if($index == $i) continue;
		return array($i, $index);	
	    }
	}
	return ;	
    }
}

$obj = new Solution();
$nums = array(3, 2, 4);
$target = 6;
$result = $obj->twoSum($nums, $target);
$result2 = $obj->twoSum2($nums, $target);
var_dump($result);
var_dump($result2);

File: com.finn/src/main/java/leetcode/TwoSum.java
package leetcode;

import java.util.HashMap;

/**
 * @author maxjoker
 * @date 2022-02-16 23:26
 *
 * 1. 两数之和
 * 给定一个整数数组 nums 和一个整数目标值 target，请你在该数组中找出 和为目标值 target  的那 两个 整数，并返回它们的数组下标。
 *
 * 你可以假设每种输入只会对应一个答案。但是，数组中同一个元素在答案里不能重复出现。
 *
 * 你可以按任意顺序返回答案。
 *
 *
 *
 * 示例 1：
 *
 * 输入：nums = [2,7,11,15], target = 9
 * 输出：[0,1]
 * 解释：因为 nums[0] + nums[1] == 9 ，返回 [0, 1] 。
 * 示例 2：
 *
 * 输入：nums = [3,2,4], target = 6
 * 输出：[1,2]
 * 示例 3：
 *
 * 输入：nums = [3,3], target = 6
 * 输出：[0,1]
 * https://leetcode-cn.com/problems/two-sum/
 */
public class TwoSum {
    /**
     * 时间复杂度：O(n^2)
     * 空间复杂度：O(n)
     * @param nums
     * @param target
     * @return
     */
    public static int[] twoSum(int[] nums, int target) {
        int[] res = new int[2];
        if (nums.length < 2) {
            return res;
        }

        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                if (nums[i] + nums[j] == target) {
                    res[0] = i;
                    res[1] = j;
                    return res;
                }
            }
        }

        return res;
    }

    /**
     * 哈希表
     * 时间复杂度：O(n)
     * 空间复杂度：O(n)
     * @param nums
     * @param target
     * @return
     */
    public static int[] twoSum2(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(target - nums[i])) {
                return new int[]{map.get(target - nums[i]), i};
            }

            map.put(nums[i], i);
        }

        return new int[0];
    }

    public static void main(String[] args) {
        int[] nums = new int[]{3,3};
        int[] res = twoSum2(nums, 6);
        for (int i : res) {
            System.out.println(i);
        }
    }
}
